contracts/Launchpad.sol

```diff
@@ -0,0 +1,492 @@
+    mapping(uint256 => uint256) public launchLinearVestingsStartTimestamps;
+    mapping(address => mapping(uint256 => uint256)) public userLinearVestingClaimTimestamp;
+    mapping(address => mapping(uint256 => mapping(uint256 => bool))) public userClaimedVestingPortion;
+
+
+    uint256 public nextLaunchId;
+    uint256 public skillPrice;
+
+    CryptoBlades _game;
+
+    function initialize(CryptoBlades game) public initializer {
+        __AccessControl_init_unchained();
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setupRole(GAME_ADMIN, msg.sender);
+
+        _game = game;
+        nextLaunchId = 1;
+    }
+
+    event LaunchAdded(uint256 indexed launchId, uint256 phase);
+    event Invested(address indexed user, uint256 amount);
+    event SkillCommitted(address indexed user, uint256 indexed launchId, uint256 amount);
+    event PeriodicVestingClaimed(address indexed user, uint256 indexed launchId, uint256 indexed vestingId, uint256 amount);
+    event LinearVestingClaimed(address indexed user, uint256 indexed launchId, uint256 amount);
+    event FundsWithdrawn(uint256 indexed launchId, uint256 amount);
+    event PeriodicVestingEnabled(uint256 indexed launchId, uint256 indexed vestingId, uint256 timestamp);
+    event LinearVestingEnabled(uint256 indexed launchId, uint256 timestamp);
+
+    // MODIFIERS
+    modifier restricted() {
+        _restricted();
+        _;
+    }
+
+    function _restricted() internal view {
+        require(hasRole(GAME_ADMIN, msg.sender), "Not game admin");
+    }
+
+    modifier onlyPhase1(uint256 launchId) {
+        _onlyPhase1(launchId);
+        _;
+    }
+
+    function _onlyPhase1(uint256 launchId) internal pure {
+        require(launchId % 2 == 1, "Usable only on phase 1");
+    }
+
+    modifier isWhitelistedFor(uint256 launchId) {
+        _isWhitelistedFor(launchId);
+        _;
+    }
+
+    function _isWhitelistedFor(uint256 launchId) internal view {
+        require(isSenderWhitelisted(launchId), "Not whitelisted");
+    }
+
+    modifier isTokenAddressSet(uint256 launchId) {
+        _isTokenAddressSet(launchId);
+        _;
+    }
+
+    function _isTokenAddressSet(uint256 launchId) internal view {
+        require(launchTokenAddress[launchId] != address(0), "Token address not set");
+    }
+
+    // VIEWS
+
+    function getTierForStakedAmount(uint256 amount) public view returns (uint256 tier) {
+        for (uint256 i = 1; i <= vars[VAR_TIERS_AMOUNT]; i++) {
+            if (amount >= tierStakingRequirement[i]) {
+                tier++;
+            } else {
+                break;
+            }
+        }
+    }
+
+    function getLaunchAllocationForTier(uint256 launchId, uint256 tier) public view returns (uint256) {
+        return launchBaseAllocation[launchId].mul(tierAllocationWeight[tier]);
+    }
+
+    function getUserRemainingAllocationForLaunch(address user, uint256 launchId) public view returns (uint256 maxAllocation) {
+        if(launchId % 2 == 0) {
```

Comment:
Wouldn't hurt to validate launch id  > 0 and < next launch unless there's no merit in the check

Reply:
I think there's no point in checking that.